### DISCUSSION TITLE
Running navidrome without macOS user login

### DISCUSSION THREAD

**Discussion by Merkwurdichliebe** [navidrome/navidrome#2335] [Q&A] [answered]
I've posted this question on the navidrome subreddit [a few weeks ago](https://www.reddit.com/r/navidrome/comments/12es5ib/running_navidrome_without_user_login/) but there were no replies.

I have navidrome installed as a service in the `LaunchAgents` folder of my separate "server" user on macOS and it's working fine. However, after battling with a weird nginx bug I found out I can run all my services (nginx, httpd) without even logging in as a macOS user, since I can simply start them from my main user via the Terminal (by using `su`).

However, it seems I can't do that with navidrome. When I try `launchctl load ~/Library/LaunchAgents/navidrome.plist` after using `su` (and without logging with the user account via the macOS GUI), I get "Load failed: 5: Input/output error".

Does the user launching the navidrome service have to be logged in as a macOS user (i.e. not via `su` Terminal) for launchctl to work, and if so, is there a way to accomplish this another way? I know I can just run navidrome in the Terminal but it ties up the command line and window. I would like to just run it and forget about it (like I can do with `nginx` and `httpd`).

Thanks a lot for your help.

**Comment by deluan** [comment, marked as answer]
It seems that you are trying to load the Navidrome service using `launchctl` from your main user account after switching to the "server" user using su. The issue you are experiencing might be due to the fact that the macOS `LaunchAgents` directory is designed to work with services associated with the user's GUI login session.

To run Navidrome without logging in as the "server" user via the macOS GUI, you can try the following approaches:

1. Use LaunchDaemons instead of LaunchAgents:
    Instead of placing the `navidrome.plist` in the LaunchAgents folder, put it in the system-wide LaunchDaemons folder: `/Library/LaunchDaemons`. LaunchDaemons run at system startup regardless of whether a user is logged in or not.

    Make sure to modify the `navidrome.plist` file accordingly to use absolute paths and proper permissions. Then, unload the current plist file:
    ```bash
    launchctl unload ~/Library/LaunchAgents/navidrome.plist
    ```
    
    Move the plist file to the LaunchDaemons folder:

    ```bash
    sudo mv ~/Library/LaunchAgents/navidrome.plist /Library/LaunchDaemons/
    ```
    Set the correct permissions:
    ```bash
    sudo chown root:wheel /Library/LaunchDaemons/navidrome.plist
    sudo chmod 644 /Library/LaunchDaemons/navidrome.plist
    ```
    
    Finally, load the plist file:
    ```bash
    sudo launchctl load /Library/LaunchDaemons/navidrome.plist
    ```

2. A "quick and dirty", kind of hacky solution would be to use `screen` (or `tmux`) to leave  Navidrome running in the background. Both are terminal multiplexers that allow you to run commands in the background, even if you close the terminal window. You can install either of them using Homebrew.

    To install and use screen:
    ```bash
    brew install screen
    su server
    screen -S navidrome
    ./path/to/navidrome
    ```

    Press Ctrl-A followed by D to detach the screen session, and you can close the terminal window.

    To reattach the session later:
    ```bash
    su server
    screen -r navidrome
    ```

    The main disadvantage of this approach is that you can't configure it to autostart after a system restart.


Either of these approaches should allow you to run Navidrome without logging in as the "server" user via the macOS GUI. Let me know if this helps.

**Reply by Merkwurdichliebe** [reply]
Thanks a lot for the detailed reply. I had indeed managed to get this to work using tmux (which I had used before with another application), but it seemed to be too hacky. The fact that you're suggesting using it makes me think it's not so terrible after all :) Thanks again.
